### PR TITLE
fix bug in shape inference of FullyConnected

### PR DESCRIPTION
**Pull request by pluskid**
ref: dmlc/mxnet#1778

### CODE REVIEW

File: src/operator/fully_connected-inl.h [hunk 1]
```diff
@@ -159,8 +159,10 @@ class FullyConnectedProp : public OperatorProperty {
     if (dshape.ndim() ==  0) return false;
 
     index_t num_input = 0;
-    mshadow::Shape<2> ishape = dshape.FlatTo2D();
-    num_input = ishape[1];
+    num_input = 1;
+    for (int i = 1; i < dshape.ndim(); ++i)
```
Comment (tqchen) [inline src/operator/fully_connected-inl.h]: dshape.ProdShape(1, dshape.ndim())

Comment (pluskid) [inline src/operator/fully_connected-inl.h]: it seems dshape is a TShape, while the ProdShape method only belongs to Shape.

Comment (tqchen) [inline src/operator/fully_connected-inl.h]: oh, I see, maybe we should add ProdShape to TShape. We can merge this in for now

Comment (tqchen) [inline src/operator/fully_connected-inl.h]: Can you add a PR to add ProdShape to TShape? I guess this could be frequently used in other places

**Comment by pluskid** [comment]
@tqchen sure I can add `ProdShape` to `TShape`. But I'm wondering maybe we should also "fix" `FlatTo2D` -- either change the doc or its implementation. Currently the doc says
```cpp
175   /*!
176    * flatten the higher dimension to second dimension, return a 2D shape
177    * \return the flat 2d shape
178    */
```
It seems the doc is indicating a different behavior from the current implementation. In the context of deep learning where the first dimension is samples, the behavior in the doc makes more sense. But I'm not sure how widely this function is used, so changing it might breaks a lot of the existing stuff. An alternative is to keep the current implementation but fix the doc.

**Comment by tqchen** [comment]
let us fix the doc, so things can be compatible

**Comment by Godricly** [comment]
I think this only solves the shape inference problem. The computation is still wrong as flat2D gives you a different shape.

**Comment by pluskid** [comment]
@Godricly Good point! Yes, it seems the computation code also uses `FlatTo2D`, and needs to be fixed.

**Comment by Godricly** [comment]
You might need to keep current version of computation for some people only want to process last dim values.

**Comment by tqchen** [comment]
@pluskid did we fix the comment @Godricly provided?

**Comment by pluskid** [comment]
@tqchen No, I have not fixed it yet. If you have a minute, can you fix it? Otherwise I will try to do it tomorrow.

**Pull request merged**